### PR TITLE
	RavenDB-6976 Allow to delete attachment via the session

### PR DESCRIPTION
**Pull request by fitzchak**
RavenDB-6976 Allow to overwrite attachment
	RavenDB-7008 Use Blake2b as the hashing function for attachments

### CODE REVIEW

File: src/Raven.Client/Documents/Session/DocumentSession.Attachments.cs [hunk 1]
```diff
@@ -73,5 +73,33 @@ private void ThrowEntityNotInSession(object entity)
...
+            if (string.IsNullOrWhiteSpace(documentId))
+                throw new ArgumentNullException(nameof(documentId));
+            if (string.IsNullOrWhiteSpace(name))
+                throw new ArgumentNullException(nameof(name));
+
+            if (_deferredCommands.OfType<DeleteCommandData>().Any(c => c.Key == documentId))
+                throw new InvalidOperationException($"Can't delete attachment {name} of document {documentId}, there is a deferred command registered for this document to be deleted.");
+
+            if (_deferredCommands.OfType<PutAttachmentCommandData>().Any(c => c.Key == documentId && c.Name == name))
+                throw new InvalidOperationException($"Can't delete attachment {name} of document {documentId}, there is a deferred command registered for this document to be deleted.");
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/DocumentSession.Attachments.cs:96]: Wrong message

File: src/Raven.Client/Documents/Session/DocumentSession.Attachments.cs [hunk 2]
```diff
@@ -73,5 +73,33 @@ private void ThrowEntityNotInSession(object entity)
...
+
+        public void DeleteAttachment(string documentId, string name)
+        {
+            if (string.IsNullOrWhiteSpace(documentId))
+                throw new ArgumentNullException(nameof(documentId));
+            if (string.IsNullOrWhiteSpace(name))
+                throw new ArgumentNullException(nameof(name));
+
+            if (_deferredCommands.OfType<DeleteCommandData>().Any(c => c.Key == documentId))
+                throw new InvalidOperationException($"Can't delete attachment {name} of document {documentId}, there is a deferred command registered for this document to be deleted.");
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/DocumentSession.Attachments.cs:93]: Why do you not allow it, isn't this a noop?

File: src/Raven.Client/Documents/Session/DocumentSession.Attachments.cs [hunk 3]
```diff
@@ -73,5 +73,33 @@ private void ThrowEntityNotInSession(object entity)
...
+
+            if (_deferredCommands.OfType<PutAttachmentCommandData>().Any(c => c.Key == documentId && c.Name == name))
+                throw new InvalidOperationException($"Can't delete attachment {name} of document {documentId}, there is a deferred command registered for this document to be deleted.");
+
+            if (DocumentsById.TryGetValue(documentId, out DocumentInfo documentInfo) &&
+                DeletedEntities.Contains(documentInfo.Entity))
+                throw new InvalidOperationException($"Can't store attachment {name} of document {documentId}, the document was already deleted in this session.");
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/DocumentSession.Attachments.cs:100]: Why, isn't this a noop?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Session/DocumentSession.Attachments.cs (hunk 1 above)
Comment (fitzchak) [reply src/Raven.Client/Documents/Session/DocumentSession.Attachments.cs:96]: Thanks, fixed.

File: src/Raven.Client/Documents/Session/DocumentSession.Attachments.cs (hunk 2 above)
Comment (fitzchak) [reply src/Raven.Client/Documents/Session/DocumentSession.Attachments.cs:93]: Yes, I change it to be no-op. Thanks!

**Pull request merged**